osbrain/agent.py

```diff
@@ -1673,36 +1686,42 @@ def run(self):
             raise
         self._running = False
         if self._kill_now:
-            # Kill the agent
-            self.kill()
+            self._die()
 
     def shutdown(self):
         """
-        Cleanly stop and shut down the agent.
-        """
-        # Stop running timers
-        self.stop_all_timers()
-        # Close all non-internal sockets
-        self.close_all()
-        # Stop the running thread
-        if self._running:
-            self.log_info('Stopping...')
-            self._keep_alive = False
-            self._kill_now = True
+        Cleanly stop and shut down the agent assuming the agent is running.
 
-    def is_running(self):
+        Will let the main thread do the tear down.
         """
-        Returns a boolean indicating whether the agent is running or not.
-        """
-        return self._running
+        self.log_info('Stopping...')
+        self._keep_alive = False
+        self._kill_now = True
 
     def kill(self):
         """
         Force shutdown of the agent.
+
+        If the agent is running the ZMQ context is terminated to allow the
+        main thread to quit and do the tear down.
+        """
+        self.stop_all_timers()
+        if self._running:
+            self._context.term()
+        else:
+            self._die()
+
+    def _die(self):
```

Comment:
The `_die` method seems overly complicated to me. I think it just adds an unnecessary amount of code. Go on with whichever solution you like best, but consider this implementation as well:
```py
def run(self):
    ...
    if self._kill_now:
        self.kill()

def kill(self):
    self.stop_all_timers()
    if self._running:
        self._context.term()
    else:
        self.close_all()
        self._pyroDaemon.shutdown()

# And delete the `die` method.
```

If you end up keeping yours, maybe consider renaming `_kill_now` to `_die_now`

Reply:
I see what you mean. I don't want to complain too much about this since it makes sense for the most part, so check out my latest revision and then we can move on with this.